compiler/src/iree/compiler/Dialect/Flow/Transforms/ConvertConv2DToWinograd.cpp

```diff
@@ -0,0 +1,281 @@
+          APFloat accum(0.0f);
+          for (int d4 = 0; d4 < kernelSize; d4++) {
+            for (int d5 = 0; d5 < kernelSize; d5++) {
+              APFloat ival(splatValue);
+              if (!isSplat) {
+                ival = input[INDEX4(d4, d5, d2, d3, shape[0], shape[1],
+                                    shape[2], shape[3])];
+              }
+              int idx0 = INDEX2(d0, d4, inputTileSize, kernelSize);
+              int idx1 = INDEX2(d1, d5, inputTileSize, kernelSize);
+              accum = accum + APFloat(G[idx0]) * ival * APFloat(G[idx1]);
+            }
+          }
+          int odx = INDEX4(d0, d1, d2, d3, inputTileSize, inputTileSize,
+                           shape[2], shape[3]);
+          output[odx] = accum;
+        }
+      }
+    }
+  }
+  return DenseElementsAttr::get(outputType, output);
+}
+
+static Value createCollapseOrExpand(
+    Value tensor, Location loc, PatternRewriter &rewriter,
+    SmallVectorImpl<int64_t> &outputShape,
+    SmallVectorImpl<ReassociationIndices> &reassociations, bool collapse) {
+  auto tensorType = tensor.getType().cast<ShapedType>();
+  auto elementTy = tensorType.getElementType();
+  auto resultType = RankedTensorType::get(outputShape, elementTy);
+  if (collapse)
+    return rewriter.create<tensor::CollapseShapeOp>(loc, resultType, tensor,
+                                                    reassociations);
+  return rewriter.create<tensor::ExpandShapeOp>(loc, resultType, tensor,
+                                                reassociations);
+}
+
+namespace {
+
+class ConvertConv2DNhwcHwcf final
+    : public OpRewritePattern<linalg::Conv2DNhwcHwcfOp> {
+ public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::Conv2DNhwcHwcfOp convOp,
+                                PatternRewriter &rewriter) const override {
+    // Check that strides = 1
+    if (!hasAllOneValues(convOp.getStrides())) return failure();
+
+    // Check that dilations = 1
+    if (!hasAllOneValues(convOp.getDilations())) return failure();
+
+    // Check that kernel size = 3x3
+    Value kernel = convOp.getInputs()[1];
+    auto kernelType = kernel.getType().cast<ShapedType>();
+    if (!kernelType) return failure();
+    ArrayRef<int64_t> kernelShape = kernelType.getShape();
+    Type elementType = kernelType.getElementType();
+    const int64_t kh = kernelShape[0];
+    const int64_t kw = kernelShape[1];
+    if ((kh != 3) || (kw != 3)) return failure();
+    const int64_t kernelSize = kh;
+
+    // TODO: Make this a user-settable parameter once we have support
+    // for more tile sizes
+    const int64_t outputTileSize = 6;
+    const int64_t inputTileSize = outputTileSize + kernelSize - 1;
+
+    // Determine if filter transform can be constant folded (only true for
+    // inference)
+    Location loc = convOp.getLoc();
+    bool constantFoldFilter{false};
+    Value foldedKernel;
+    auto constOp = kernel.getDefiningOp<arith::ConstantOp>();
```

Comment:
Use `matchPattern(kernel, m_Constant)` here to avoid pinning to `arith::Constant` specifically?

Reply:
Nice, I haven't used this function before but seems much better. Will switch to using this.